### PR TITLE
cons: Show last line of output even if it has no `\n`

### PR DESCRIPTION
**Pull request by kazarmy**
**Your checklist for this pull request**
- [X] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [X] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [ ] I've documented or updated the documentation of every function and struct this PR changes. If not so I've explained why.
- [X] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

This pr makes the rz_cons prompt show the last line of command output even if it has no '`\n`' (newline). The following before/after images illustrate this.

From `rizin bins/bf/2+5.bf`:

__Before:__

before-nonl_last_line

__After:__

after-nonl_last_line

For this to work with `analysis_bf.c`, it needs to call into rz_cons (currently indirectly using `cb_printf`) instead of calling `putchar()` directly.

**Test plan**

All builds are green.

**Closing issues**

...

### CODE REVIEW

File: librz/cons/cons.c [hunk 1]
```diff
@@ -1023,6 +1025,18 @@ RZ_API void rz_cons_flush(void) {
...
+		const char *p = &I.context->buffer[I.context->buffer_len - 1];
+		for (; *p != '\n' && p > I.context->buffer; p--)
+			;
+		if (*p == '\n') {
+			RZ_FREE(I.context->last_nonl_line);
+			p++;
+		}
+		I.context->last_nonl_line = rz_str_append(I.context->last_nonl_line, p);
+	}
+
```
Comment (ret2libc) [inline librz/cons/cons.c:1039]: Is this really necessary? Can't we force the output by adding a `rz_cons_printf("\n");` in the `aezsu` command? I think it is better than having rz_cons automagically perform this.

File: librz/analysis/p/analysis_bf.c [hunk 2]
```diff
@@ -36,6 +36,8 @@ static void bf_syscall_read(RzILVM *vm, RzILOpEffect *op) {
 	rz_bv_free(bv);
 }
 
+static PrintfCallback cb_printf = NULL;
```
Comment (wargio) [inline librz/analysis/p/analysis_bf.c:39]: i really do not like this.

**Comment by wargio** [comment]
i have a counter proposal, what if we implement stdout via RzStrBuf ? or even better, how about just read/write from some memory spaces? like add a fake register and have these to hold a circular buffer that is used as stdin/out ?

Thse fake registers can be simpply called stdin and stdout.

File: librz/cons/cons.c (hunk 1 above)
Comment (kazarmy) [reply librz/cons/cons.c:1039]: > Is this really necessary?

Strictly speaking, for `aeszu` yes an extra newline is just an inaccuracy, but suppose a buggy command produces output that is not terminated with a newline. The current code will just make the final line disappear and you'd be lucky if you realize what was happening.

> Can't we force the output by adding a `rz_cons_printf("\n");` in the `aezsu` command?

Yes but if the emulated code does actually print a newline then there will be a double newline. The only way `aezsu` can determine whether a newline is needed or not is to check what the final character of the output is, and for that you need code in rz_cons that is probably similar to the above.

Comment (ret2libc) [reply librz/cons/cons.c:1039]: > Yes but if the emulated code does actually print a newline then there will be a double newline.

So be it. If you are emulating a new line, than it must print the newline and that should not have anything to do with the additional newline that we print by defaul imo.

File: librz/analysis/p/analysis_bf.c (hunk 2 above)
Comment (kazarmy) [reply librz/analysis/p/analysis_bf.c:39]: it is decreed that all rizin output must go through rz_cons but i see your comment below

File: librz/cons/cons.c (hunk 1 above)
Comment (kazarmy) [reply librz/cons/cons.c:1039]: Besides, the bash shell works like this and I'm just following bash in this regard.

**Comment by kazarmy** [comment]
> what if we implement stdout via RzStrBuf ?

rz_cons isn't a string buffer?

> or even better, how about just read/write from some memory spaces? like add a fake register and have these to hold a circular buffer that is used as stdin/out ?

this should be done as well, but if you wanted to see output on-the-fly ... but I suppose output like that should be printed on a different terminal other than the rizin one

File: librz/cons/cons.c (hunk 1 above)
Comment (ret2libc) [reply librz/cons/cons.c:1039]: > Besides, the bash shell works like this

What do you mean?

```
ret2libc@fedora:~$ echo -n "Hello"
Helloret2libc@fedora:~$ echo -ne "Hello\n\n"
Hello

ret2libc@fedora:~$ 
```

Comment (kazarmy) [reply librz/cons/cons.c:1039]: Exactly!

> So be it.

Ok.

**Comment by kazarmy** [comment]
Closed to pursue alternative similar to https://github.com/rizinorg/rizin/pull/2385#discussion_r820618089.

**Comment by kazarmy** [comment]
Or maybe the same as https://github.com/rizinorg/rizin/pull/2385#discussion_r820618089.

**Pull request closed without merging**